numba/targets/arraymath.py

```diff
@@ -399,22 +399,52 @@ def array_cumprod_impl(arr):
     return impl_ret_new_ref(context, builder, sig.return_type, res)
 
 
-@lower_builtin(np.mean, types.Array)
-@lower_builtin("array.mean", types.Array)
-def array_mean(context, builder, sig, args):
-    zero = sig.return_type(0)
+@register_jitable
+def sum_array(arr, dtype):
+    return np.sum(arr, dtype=dtype)
 
-    def array_mean_impl(arr):
-        # Can't use the naive `arr.sum() / arr.size`, as it would return
-        # a wrong result on integer sum overflow.
-        c = zero
-        for v in np.nditer(arr):
-            c += v.item()
-        return c / arr.size
 
-    res = context.compile_internal(builder, array_mean_impl, sig, args,
-                                   locals=dict(c=sig.return_type))
-    return impl_ret_untracked(context, builder, sig.return_type, res)
+@register_jitable
+def sum_array_axis(arr, axis, dtype):
+    return np.sum(arr, axis=axis, dtype=dtype)
+
+
+@overload(np.mean)
+@overload_method(types.Array, 'mean')
+def array_mean(arr, axis=None):
+    if isinstance(arr, types.Array):
+        # determine accumulator type
+        if isinstance(arr.dtype, (types.Integer, types.Boolean)):
+            ret_dtype = np.float64
+        elif isinstance(arr.dtype, (types.Float, types.Complex)):
+            ret_dtype = arr.dtype
+        else:
+            raise TypeError(("np.mean is not supported on {} arrays. "
+                            "It supports boolean, integer, float "
+                            "and complex arrays").format(arr.dtype))
+        # dispatch based on whether there's an axis parameter and its type
+        if axis is None:
```

Comment:
```suggestion
        if axis is None or isinstance(axis, types.Omitted):
```
will get rid of the error.  It is part of a weird Numba behavior.  See https://github.com/numba/numba/issues/4694